src/hotspot/cpu/riscv/macroAssembler_riscv.cpp

```diff
@@ -2605,6 +2605,47 @@ void MacroAssembler::movptr2(Register Rd, uint64_t addr, int32_t &offset, Regist
   offset = lower12;
 }
 
+// floating point imm move
+bool MacroAssembler::can_fp_imm_load(float imm) {
+  jint f_bits = jint_cast(imm);
+  if (f_bits == 0) {
+    return true;
+  }
+  int dummy;
+  return can_zfa_zli_float(imm, &dummy);
+}
+
+bool MacroAssembler::can_dp_imm_load(double imm) {
+  julong d_bits = julong_cast(imm);
+  if (d_bits == 0) {
+    return true;
+  }
+  int dummy;
+  return can_zfa_zli_double(imm, &dummy);
+}
+
+void MacroAssembler::fli_s(FloatRegister Rd, float imm) {
+  jint f_bits = jint_cast(imm);
+  if (f_bits == 0) {
+    fmv_w_x(Rd, zr);
+    return;
+  }
+  int Rs = -1;
+  can_zfa_zli_float(imm, &Rs);
```

Comment:
should that be a `guarantee(can_zfa_zli_float(imm, &Rs));`?

Reply:
Same, my bad. (guarantee in zli_s on that Rs is uimm5.)

As we already an guarantee, maybe assert(can_zfa..) ?

Reply:
How about we do this directly on the callsite?
```
  int Rs = -1;
  if (can_zfa_zli_float(imm, &Rs)) {
    _fli_s(Rd, Rs);
  }
```

Reply:
Or can we use `zfa_zli_lookup_float/double` instead?

Reply:
Then we loose the assert on UseZfa and we still need to cast to bits, so basically copy out the code from can_zfa_zli_float.

I'll merge with zfh and try to address your concerns.